harness/tests/determined/pytorch/experimental/test_torch_batch_process.py

```diff
@@ -0,0 +1,570 @@
+    get_default_device,
+    torch_batch_process,
+)
+
+DEFAULT_SLOT_IDS = [0]
+DEFAULT_ADDRS = ["0.0.0.12"]
+DEFAULT_DEVICE = torch.device("cuda", 1)
+DEFAULT_STORAGE_PATH = "default_storage_path"
+
+
+@pytest.fixture
+def batch_processor_with_avg_metric_reducer() -> Type[TorchBatchProcessor]:
+    class MySumMetricReducer(pytorch.MetricReducer):
+        def __init__(self) -> None:
+            self.reset()
+
+        def reset(self) -> None:
+            self.sum = 0
+
+        def update(self, value: List[int]) -> None:
+            self.sum += sum(value)
+
+        def per_slot_reduce(self) -> int:
+            return self.sum
+
+        def cross_slot_reduce(self, per_slot_metrics: List[int]) -> int:
+            return sum(per_slot_metrics)
+
+    class MyProcessor(TorchBatchProcessor):
+        def __init__(self, context: TorchBatchProcessorContext) -> None:
+            self.context = context
+            self.reducer_sum = context.wrap_reducer(reducer=MySumMetricReducer(), name="sum_metric")
+
+        def process_batch(self, batch: Any, batch_idx: int) -> None:
+            self.reducer_sum.update(batch)
+
+    return MyProcessor
+
+
+def _get_index_dataset(data_length: int = 50) -> Dataset:
+    class IndexData(Dataset):
+        def __init__(self, data_length: int) -> None:
+            self.data = data_length
+
+        def __len__(self) -> int:
+            return int(self.data)
+
+        def __getitem__(self, idx: int) -> int:
+            return idx
+
+    return IndexData(data_length)
+
+
+def _get_core_context(
+    rank: int = 0, should_preempt_results: Optional[List[bool]] = None
+) -> MagicMock:
+    mock_core_context = MagicMock()
+    mock_distributed_context = MagicMock()
+    mock_distributed_context.get_rank.return_value = rank
+    mock_distributed_context.broadcast.return_value = "mock_checkpoint_uuid"
+    mock_core_context.__enter__().distributed = mock_distributed_context
+    if should_preempt_results is None:
+        mock_core_context.__enter__().preempt.should_preempt.return_value = False
+    else:
+        mock_core_context.__enter__().preempt.should_preempt.side_effect = should_preempt_results
+    return mock_core_context
+
+
+def _get_det_info(
+    slot_ids: List[int], container_addrs: List[str], latest_checkpoint: Optional[str] = None
+) -> MagicMock:
+    mock_cluster_info = MagicMock()
+    mock_cluster_info.slot_ids = slot_ids
+    mock_cluster_info.container_addrs = container_addrs
+    mock_cluster_info.latest_checkpoint = latest_checkpoint
+    return mock_cluster_info
+
+
+@patch("determined.pytorch.experimental._torch_batch_process._initialize_default_inference_context")
+@patch("determined.get_cluster_info")
+@patch("determined.pytorch.experimental._torch_batch_process._synchronize_and_checkpoint")
+@patch("determined.pytorch.experimental._torch_batch_process._report_progress_to_master")
+@patch("determined.pytorch.experimental._torch_batch_process._reduce_metrics")
```

Comment:
This feels like an alarming amount of patching.

I wonder if perhaps its possible to break up the monster `torch_batch_process` into components, and test each one separately.  Can you make it so you don't have to test so much stuff, just to test input validation?

Reply:
Reduced # of patches with your comment, in some cases, patches is still needed to assert against the mock.

Also, for simple things like kawargs validation, directly tested against the private function.